### PR TITLE
 Optimize the support for virtual keyboard

### PR DESCRIPTION
**Pull request by kingysu**
1. There are two kinds of virtual keyboard at least.
    
2. The input method engine may support one or two function mode: full or limited.
   a) The full function mode is the default function mode. All the input method engine
   which does not implement InputMethodEngineV4 or its ascendants supports the full
   function mode.
   b) The limited function mode is a dedicated function mode which is used for the true
   virtual keyboard. The input method engine which implements InputMethodEngineV4 or its
   ascendants should tell fcitx which function mode it supports. That is, it may support
   full function mode or limited function mode or both.
    
3. The two different virtual keyboard use different function mode of input method engine.
   The virtual keyboard should tell fcitx which function mode it needs. Fcitx will decide
   which function mode is used. For example, if the virtual keyboard tells fcitx that it
   needs limited function mode, but the input method engine does not support that mode,
   then fcitx will use full function mode for compitibility reason.

### CODE REVIEW

**Comment by wengxt** [comment]
Question, should function mode be a capability flag instead ? if capability flag bits are limited, we can introduce a new property for input context. Associate it to every key event seems wrong to me.

**Comment by kingysu** [comment]
> Question, should function mode be a capability flag instead ? if capability flag bits are limited, we can introduce a new property for input context. Associate it to every key event seems wrong to me.

I have updated the PR to make sure that the virtual keyboard can nofity fcitx5 of its function mode.

**Review by wengxt — CHANGES_REQUESTED** [review]

File: src/lib/fcitx/inputcontext_p.h [hunk 1]
```diff
@@ -132,6 +133,14 @@ class InputContextPrivate : public QPtrHolder<InputContext> {
 
     InputContextProperty *property(int slot) { return properties_[slot].get(); }
 
+    bool isFunctionModeSupported(FunctionMode functionMode) {
+        FCITX_Q();
+
+        auto *instance = manager_.instance();
+        auto *engine = instance->inputMethodEngine(q);
```
Comment (wengxt) [inline src/lib/fcitx/inputcontext_p.h:140]: engine can be null, even instance can be null.

File: src/lib/fcitx/inputmethodengine.cpp [hunk 2]
```diff
@@ -28,18 +27,43 @@ std::string InputMethodEngine::subModeLabel(const InputMethodEntry &entry,
...
+        FCITX_WARN() << "InputMethodEngine::virtualKeyboardEvent is called."
+                     << "But there is no valid InputMethodEngineV4 instance."
+                     << "So it returns early.";
+        return;
     }
+
+    this4->virtualKeyboardEventImpl(entry, virtualKeyboardEvent);
+}
+
+FunctionMode InputMethodEngine::getFunctionMode() const {
```
Comment (wengxt) [inline src/lib/fcitx/inputmethodengine.cpp:41]: I don't quite understand why function mode needs a setter function. shouldn't it be a static property?

Comment (wengxt) [reply src/lib/fcitx/inputmethodengine.cpp:41]: It seems that the mode is a instance property, then shouldn't engine just read this property?

**Review by wengxt — CHANGES_REQUESTED** [review]

File: src/lib/fcitx/inputcontext.h [hunk 3]
```diff
@@ -41,6 +41,8 @@ class InputPanel;
 class StatusArea;
 typedef std::function<bool(InputContext *ic)> InputContextVisitor;
 
+enum class FunctionMode : uint32_t { Full = 1 << 0, Limited = 1 << 1 };
```
[the remark this answers is not in the document]
Comment (kingysu) [reply src/lib/fcitx/inputcontext.h:44]: OK. I will fix it.

File: src/lib/fcitx/inputcontext_p.h (hunk 1 above)
Comment (kingysu) [reply src/lib/fcitx/inputcontext_p.h:140]: OK. I will fix it.

File: src/lib/fcitx/inputmethodengine.cpp (hunk 2 above)
Comment (kingysu) [reply src/lib/fcitx/inputmethodengine.cpp:41]: The input method engine can support just full function mode or support both full function mode and limited funciton. So we should tell the input method engine which function mode should be effective. Or else, the input method engine should use full function mode when keyEvent function is caled and limited function mode when virtualKeyboardEvent function is called.

**Comment by kingysu** [comment]
> Question, should function mode be a capability flag instead ? if capability flag bits are limited, we can introduce a new property for input context. Associate it to every key event seems wrong to me.

done

File: src/ui/virtualkeyboard/virtualkeyboard.cpp [hunk 4]
```diff
@@ -120,6 +125,16 @@ class VirtualKeyboardBackend
     VirtualKeyboard *parent_;
 };
 
+void VirtualKeyboardBackend::setVirtualKeyboardFunctionMode(uint32_t mode) {
+    if (mode != static_cast<uint32_t>(VirtualKeyboardFunctionMode::Full) &&
+        mode != static_cast<uint32_t>(VirtualKeyboardFunctionMode::Limited)) {
+        return;
```
Comment (wengxt) [inline src/ui/virtualkeyboard/virtualkeyboard.cpp:131]: throw invalid argument error here, 

See:
https://github.com/fcitx/fcitx5/blob/00d52a4a34aaabd2fac922deff1860c5ef04e759/src/modules/dbus/dbusmodule.cpp#L295

**Review by wengxt — CHANGES_REQUESTED** [review]

File: src/ui/virtualkeyboard/virtualkeyboard.cpp (hunk 4 above)
Comment (kingysu) [reply src/ui/virtualkeyboard/virtualkeyboard.cpp]: fix it.

Comment (kingysu) [reply src/ui/virtualkeyboard/virtualkeyboard.cpp]: > throw invalid argument error here,
> 
> See:
> 
> https://github.com/fcitx/fcitx5/blob/00d52a4a34aaabd2fac922deff1860c5ef04e759/src/modules/dbus/dbusmodule.cpp#L295

fix it.

**Pull request merged**